src/Akka.Persistence.Redis/Query/Stages/AllEventsSource.cs

```diff
@@ -0,0 +1,343 @@
+                {
+                    switch (_state)
+                    {
+                        case State.Initializing:
+                            _state = State.QueryWhenInitializing;
+                            break;
+                        default:
+                            Query();
+                            break;
+                    }
+                });
+            }
+
+            public override void PreStart()
+            {
+                _callback = GetAsyncCallback<(int, IReadOnlyList<(string, IPersistentRepresentation)>)>(raw =>
+                {
+                    var nb = raw.Item1;
+                    var events = raw.Item2;
+
+                    if (events.Count == 0)
+                    {
+                        if (_currentOffset >= _maxOffset)
+                        {
+                            // end has been reached
+                            CompleteStage();
+                        }
+                        else
+                        {
+                            switch (_state)
+                            {
+                                case State.NotifiedWhenQuerying:
+                                    // maybe we missed some new event when querying, retry
+                                    _state = State.Idle;
+                                    Query();
+                                    break;
+                                case State.Querying:
+                                    if (_live)
+                                    {
+                                        // nothing new, wait for notification
+                                        _state = State.WaitingForNotification;
+                                    }
+                                    else
+                                    {
+                                        // not a live stream, nothing else currently in the database, close the stream
+                                        CompleteStage();
+                                    }
+                                    break;
+                                default:
+                                    Log.Error($"Unexpected source state: {_state}");
+                                    FailStage(new IllegalStateException($"Unexpected source state: {_state}"));
+                                    break;
+                            }
+                        }
+                    }
+                    else
+                    {
+                        var evts = events.ZipWithIndex().Select(c =>
+                        {
+                            var repr = c.Key.Item2;
+                            if (repr != null && !repr.IsDeleted)
+                            {
+                                return new EventEnvelope(new Sequence(_currentOffset + c.Value), repr.PersistenceId, repr.SequenceNr, repr.Payload);
+                            }
+
+                            return null;
```

Comment:
Does this suffice? 
````csharp
var evts = events.ZipWithIndex().Select(c =>
                        {
                            var repr = c.Key.Item2;
                            if (repr != null && !repr.IsDeleted)
                            {
                                return new EventEnvelope(new Sequence(_currentOffset + c.Value), repr.PersistenceId, repr.SequenceNr, repr.Payload);
                            }

                            return null;
                        })//in case of null
                        .Where(x=> x != null).ToList();
````

Reply:
It makes more sense to have `Where` before `Select`

Reply:
The `Where` is dealing with the null that maybe returned from the `Select`.

Reply:
you can do a `.Where(kvp => kvp.Key.Item2 != null && !kvp.Key.Item2.IsDeleted)`

Reply:
I will go with that. 
Thanks!